Source/WebCore/style/ElementRuleCollector.cpp

```diff
@@ -584,14 +584,65 @@ bool ElementRuleCollector::containerQueriesMatch(const RuleData& ruleData, const
 
 bool ElementRuleCollector::scopeRulesMatch(const RuleData& ruleData, const MatchRequest& matchRequest)
 {
-    auto queries = matchRequest.ruleSet.scopeRulesFor(ruleData);
+    auto rules = matchRequest.ruleSet.scopeRulesFor(ruleData);
 
-    if (queries.isEmpty())
+    if (rules.isEmpty())
         return true;
 
-    // FIXME: to implement
-    // https://bugs.webkit.org/show_bug.cgi?id=265241
-    return false;
+    SelectorChecker checker(element().rootElement()->document());
+    auto context = SelectorChecker::CheckingContext(SelectorChecker::Mode::ResolvingStyle);
+
+    auto evaluate = [&](auto& rule) {
+        Element* olderAncestorToMatch = nullptr;
+
+        auto ancestorsMatch = [&](auto& selector) {
+            auto ancestor = element().parentElement();
+            while (ancestor) {
+                if (ancestor == olderAncestorToMatch) {
+                    // We limit the search to find an end of the donut to children of the start of the donut, not an ancestor
+                    return false;
+                }
+                auto match = checker.match(selector, *ancestor, context);
+                if (match) {
+                    olderAncestorToMatch = ancestor;
+                    return true;
+                }
+                ancestor = ancestor->parentElement();
+            }
+            return false;
+        };
+
+        auto scopeStart = rule->scopeStart();
+        if (!scopeStart.isEmpty()) {
+            CSSSelector selector;
+            selector.setMatch(CSSSelector::Match::PseudoClass);
+            selector.setPseudoClassType(CSSSelector::PseudoClassType::Is);
+            selector.setSelectorList(makeUnique<CSSSelectorList>(WTFMove(scopeStart)));
+            if (!ancestorsMatch(selector))
+                return false;
+        }
```

Comment:
Copying and constructing selectors in middle of style resolution is surely not the way to go? Why can't we match the scopeStart() selector list directly?

Reply:
The behaviour described in spec is purposely similar to `:is()`  (ignore pseudo-element) and `matchSelectorList` is private in `SelectorChecker` and needs a `LocalContext` so it seems more direct to wrap it in `:is`. 

Should we try to adjust `SelectorChecker::matchSelectorList` a bit for this ?

Reply:
You can just loop over the selectors and call match(). One of the existing SelectorChecker modes likely gives you the right behavior and if not you can add a new one.